crates/humanode-peer/src/service/inherents.rs

```diff
@@ -104,3 +104,39 @@ impl<Client> Clone for Creator<Client> {
         }
     }
 }
+
+/// Sealing case related inherent data providers.
+pub mod sealing {
+    use std::cell::RefCell;
+
+    thread_local!(static TIMESTAMP: RefCell<u64> = RefCell::new(0));
+
+    /// Provide a simulated duration starting at 0 in millisecond for timestamp inherent.
+    /// Each call will increment timestamp by slot duration making Babe think time has passed.
+    pub struct SimulatedTimestampInherentDataProvider;
+
+    #[async_trait::async_trait]
+    impl sp_inherents::InherentDataProvider for SimulatedTimestampInherentDataProvider {
+        async fn provide_inherent_data(
+            &self,
+            inherent_data: &mut sp_inherents::InherentData,
+        ) -> Result<(), sp_inherents::Error> {
+            TIMESTAMP.with(|x| {
+                let current_timestamp = *x.borrow();
+                *x.borrow_mut() = current_timestamp
+                    .checked_add(humanode_runtime::SLOT_DURATION)
+                    .expect("operation is under control, sealing case is allowed only in dev mode");
+                inherent_data.put_data(sp_timestamp::INHERENT_IDENTIFIER, &*x.borrow())
+            })
```

Comment:
This code assumes `provide_inherent_data` is called once for each block. I am not sure the code we have makes any effort to guarantee this - so no reason why this should be true. It seems like we need to do something else here.